### DISCUSSION TITLE
Difficulty using Located with parser nesting

### DISCUSSION THREAD

**Discussion by Techcable** [winnow-rs/winnow#572] [Q&A]
I am having trouble tracking spans with `Located` and nested input.

I have two parsers, an inner parser & outer parser.

I want to parse till the end of the line in the outer parser, then pass it to the inner parser.

First I call `winnow::ascii::till_line_ending` which is a ` impl Parser<Located<&str>, &str, _>`
To try to recover the input location i call `take` which is also an ` impl Parser<Located<&str>, &str, _>`

But then I can't call `and_then(inner_parser)` because the inner parser expects location information that has been discarded.

As far as I can tell, there is no way for me to "nest" the inner parser inside the outer parser without loosing location information.

The issue is that both `till_line_ending` and `take` produce `<I as Stream>::Slice` while I want the original input `I`.

```rust
fn inner_parser<'a>() -> impl Parser<Located<&'a str>, Inner, ContextError> {
    ...
}

fn outer_parser<'a>()  -> impl Parser<Located<&'a str>, Outer, ContextError>  {
    seq!(
        winnow::ascii::till_line_ending // impl Parser<Located<&str>, &str, _>
            .take() // impl Parser<Located<&str>, &str, _>
            .and_then(
                // MISMATCHED TYPES!
                // expected type: impl Parser<Located<&str>, Inner, _>
                // actual type: impl Parser<&str, Inner, _>
                inner_parser()
            ),
        winnow::ascii::till_line_ending
            .map(String::from)
    )
    .with_span()
    .map(|((first, second), span)| Outer { first, second, span })
}
```

Complete code & compiler errors are here: https://gist.github.com/Techcable/be3a473722bed761599d80186ba79f8f

For now I've worked around this by simply discarding span location.
However, that is obviously suboptimal.

**Comment by epage** [comment]
Feel free to make an issue for improving this!

As a workaround, you could
- clone the outer input
- [`Stream::reset` the clone of the outer input to the [`Stream::checkpoint`] from the inner input
- parse that new inner input

This will likely require dropping down to imperative code.
